emasamplerate.go

```diff
@@ -0,0 +1,297 @@
+package dynsampler
+
+import (
+	"math"
+	"sort"
+	"sync"
+	"time"
+)
+
+// EMASampleRate implements Sampler and attempts to average a given sample rate,
+// weighting rare traffic and frequent traffic differently so as to end up with
+// the correct average. This method breaks down when total traffic is low
+// because it will be excessively sampled.
+//
+// Based on the AvgSampleRate implementation, EMASampleRate differs in that rather
+// than compute rate based on a periodic sample of traffic, it maintains an Exponential
+// Moving Average of counts seen per key, and adjusts this average at regular intervals.
+// The weight applied to more recent intervals is defined by `weight`, a number between
+// (0, 1) - larger values weight the average more toward recent observations.
+//
+// Keys that are not found in the EMA will always have a sample
+// rate of 1. Keys that occur more frequently will be sampled on a logarithmic
+// curve. In other words, every key will be represented at least once in any
+// given window and more frequent keys will have their sample rate
+// increased proportionally to wind up with the goal sample rate.
+type EMASampleRate struct {
+	// AdjustmentInterval defines how often we adjust the moving average from recent observations
+	// Default 15s
+	AdjustmentInterval int
+
+	// Weight is a value between (0, 1) indicating the weighting factor used to adjust
+	// the EMA. With larger values, newer data will influence the average more, and older
+	// values will be factored out more quickly. Default is 0.5
+	Weight float64
+
+	// GoalSampleRate is the average sample rate we're aiming for, across all
+	// events. Default 10
+	GoalSampleRate int
+
+	// MaxKeys, if greater than 0, limits the number of distinct keys tracked in EMA.
+	// Once MaxKeys is reached, new keys will not be included in the sample rate map, but
+	// existing keys will continue to be be counted.
+	MaxKeys int
+
+	// AgeOutValue indicates the threshold for removing keys from the EMA.
+	// The EMA of any key will approach 0 if it is not repeatedly observed, but will never truly reach it, so
+	// we have to decide what constitutes "zero".
+	// Keys with averages below this threshold will be removed from the EMA. Default is the same as Weight
+	AgeOutValue float64
+
+	// BurstMultiple, if set, is multiplied by the sum of the running average of counts to define the burst detection threshold.
+	// If total counts observed for a given interval exceed the threshold, EMA is updated immediately, rather than waiting on the
+	// AdjustmentInterval. This throughput-based adjustment allows the dynamic sample rates to reflect changes in traffic from
+	// short bursts. Defaults to 2, negative value disables
+	BurstMultiple float64
+
+	// BurstDetectionDelay indicates the number of intervals to run before burst detection kicks in.
+	// Defaults to 3
+	BurstDetectionDelay uint
```

Comment:
I don't understand this one. I though burst detection meant re-evaluating the EMA would happen in _less than_ an interval, but this makes me think it will take _several intervals_ to take effect. In which case.... why is it here?

Reply:
It's so that when we start up a new instance of the sampler, we don't go immediately into burst mode because the EMA avg count will be at zero and 2 x anything will be > zero.

Reply:
Changed this to: